core/eval.go

```diff
@@ -860,6 +862,351 @@ func evalQWATCH(args []string, c *Client) []byte {
 	return RESP_OK
 }
 
+// SETBIT key offset value
+func evalSETBIT(args []string) []byte {
+	var err error
+
+	if len(args) != 3 {
+		return Encode(errors.New("ERR wrong number of arguments for 'setbit' command"), false)
+	}
+
+	key := args[0]
+	offset, err := strconv.ParseInt(args[1], 10, 64)
+	if err != nil {
+		return Encode(errors.New("ERR bit offset is not an integer or out of range"), false)
+	}
+
+	value, err := strconv.ParseBool(args[2])
+	if err != nil {
+		return Encode(errors.New("ERR bit is not an integer or out of range"), false)
+	}
+
+	obj := Get(key)
+	requiredByteArraySize := offset/8 + 1
+
+	if obj == nil {
+		obj = NewObj(NewByteArray(int(requiredByteArraySize)), -1, OBJ_TYPE_BYTEARRAY, OBJ_ENCODING_BYTEARRAY)
+		Put(args[0], obj)
+	}
+
+	// handle the case when it is string
+	if assertType(obj.TypeEncoding, OBJ_TYPE_STRING) == nil {
+		return Encode(errors.New("ERR value is not a valid byte array"), false)
+	}
+
+	// handle the case when it is byte array
+	if assertType(obj.TypeEncoding, OBJ_TYPE_BYTEARRAY) == nil {
+		byteArray := obj.Value.(*ByteArray)
+		byteArrayLength := byteArray.Length
+
+		// check whether resize required or not
+		if requiredByteArraySize > byteArrayLength {
+			// resize as per the offset
+			byteArray = byteArray.IncreaseSize(int(requiredByteArraySize))
+		}
+
+		response := byteArray.GetBitByByteArrayPosition(int(offset))
+		byteArray.SetBitByByteArrayPosition(int(offset), value)
+
+		// if earlier bit was 1 and the new bit is 0
+		// propability is that, we can remove some space from the byte array
```

Comment:
Is this an optimisation to save memory when a bit has been unset? Is this consistent with redis?